File: python/z1.py
def recurs(a, chisla, s):
    if len(chisla) == 0:
        return a == s
    for i in range(len(chisla)):
        oleg = []
        deleted = False
        for j in range(len(chisla)):
            if chisla[j] != chisla[i] or deleted:
                oleg.append(chisla[j])
            else:
                deleted = True
        if recurs(a + chisla[i], oleg, s):
            print(chisla[i], "+", end=" ")
            return True
        elif recurs(a * chisla[i], oleg, s):
            print(chisla[i], "*", end=" ")
            return True
    return False

def start(chisla, s):
    for i in range(len(chisla)):
        oleg = []
        deleted = False
        for j in range(len(chisla)):
            if chisla[j] != chisla[i] or deleted:
                oleg.append(chisla[j])
            else:
                deleted = True
        if recurs(chisla[i], oleg, s):
            print(chisla[i])
            return True
    return False

def main():
    s = int(input())
    chisla = [89, 65, 56, 43]
    print()
    print(start(chisla, s))

if __name__ == "__main__":
    main()

File: cpp/z1.cpp
#include <iostream>
#include <vector>

using namespace std;

bool recurs(int a, vector<int> chisla, const int& s) {
	if (chisla.size() == 0) {
		return (a == s);
	}
	for (size_t i = 0; i < chisla.size(); i++) {
		vector<int> oleg = {};
		bool deleted = false;
		for (size_t j = 0; j < chisla.size(); j++){
			if (chisla[j] != chisla[i] || deleted) {
				oleg.push_back(chisla[j]);
			}
			else deleted = true;
		}

		if (recurs(a + chisla[i], oleg, s)) {
			cout << chisla[i] << " + " ;
			return true; 
		}
		else if (recurs(a * chisla[i], oleg, s)) {
			cout << chisla[i] << " * ";
			return true;
		}
	}
	return false;
}

bool start(vector<int> chisla, const int& s) {
	for (size_t i = 0; i < chisla.size(); i++) {
		vector<int> oleg = {};
		bool deleted = false;
		for (size_t j = 0; j < chisla.size(); j++) {
			if (chisla[j] != chisla[i] || deleted) {
				oleg.push_back(chisla[j]);
			}
			else deleted = true;
		}

		if (recurs(chisla[i], oleg, s)) {
			cout << chisla[i];
			return true;
		}
	}
	return false;
}

int main() {
	int s;
	int n = 4;

	vector<int> chisla = { 89, 65, 56, 43 };

	cin >> s;

	cout << endl << start(chisla, s);
}